keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, this requires that the users of this feature belong to the cluster network. TODO define a way of filtering the headers in the answer, per example, by source IP, certificate name, ... 
+
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
+```
+
+The Alternative Services RFC defines options to [implement caching](https://datatracker.ietf.org/doc/html/rfc7838#section-3.1),
+however, it doesn't seem necessary because:
+
+- kubernetes control-plane is not a high dynamic environment.
+- apiserver already has another mechanisms to warn clients using http codes.
+- if the alt-service has gone you will have a connection error, the client will expire that entry from the cache and will try another alt-svc.
+
+### Client side HA
+
+Client-go creates a layer on top of the golang net/http library abstracting the communications against the apiservers using [Requests](https://github.com/kubernetes/client-go/blob/master/rest/request.go).
+
+The golang net/http library allows to modify a request in different points:
+- RoundTripper/Transport
+- Dialer
+- DNS dialer
+
+The client-go base-URL is immutable after creation, that guarantees that Requests will not be able to modify it externally.
+
+Client-go already implements some [custom RoundTrippers](https://github.com/kubernetes/client-go/blob/master/transport/round_trippers.go) for some functionalities, like debugging, authentication, impersonation, ...
+
+The proposal is to implement a new round tripper that has a local cache with the list of available apiservers.
+This list can be created manually via configuration (comma-separated list on the Config object) or automatically via the Alt-Svc headers published.
+
+The round tripper behavior, if there are alternative services, will be:
+
+1. Prefer local (same system as the client) alternative services.
+2. Prefer the one in the original request if is listed as an alternative service.
+3. If one alternative service is chosen and succeed, stick to it, only change on network failures.
```

Comment:
I think the rule should be either round robin or randomly select one of the available endpoints (and remove an endpoint from consideration for X minutes if it turns out to be unreachable).

(Meaning, delete step 1 and 2 in this list-- I am not sure they are a good thing)

Reply:
point 2 is debatable, but having a local connection you can assume latency close to 0 and practically infinite buffer with close to 0 chances of packet drops, your network using local is a buffer in the kernel. 
Using connections through the network, they are limited by the NICs and underlay network, and exposed to all the problem of those

Reply:
Clients that want HA don't care about that (note apiserver could be down).

Clients that want low latency/speed should not need to use this mechanism at all.

Reply:
Now that you put it that way it may be a niche use case ... I'm happy to drop it

Reply:
In  general we try to steer designs away from optimizing for locality.  It quickly stops being an optimization and becomes a functional requirement.